test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards.NodeGatewayStartedShards;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.store.TransportNodesListShardStoreMetaData.NodeStoreFilesMetaData;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A gateway allocator implementation that keeps an in memory list of started shard allocation
+ * that are used as replies to the, normally async, fetch data requests.
+ *
+ * For now only primary shard related data is fetched. Replica request always get an empty response.
+ *
+ * This class is useful to use in unit tests that require the functionality of {@link GatewayAllocator} but do
+ * not have all the infrastructure required to use it.
+ */
+public class TestGatewayAllocator extends GatewayAllocator {
+
+    Map<String, Map<ShardId,ShardRouting>> knownAllocations = new HashMap<>();
+    Map<String, DiscoveryNode> knownNodes = new HashMap<>();
+
+    PrimaryShardAllocator primaryShardAllocator = new PrimaryShardAllocator(Settings.EMPTY) {
+        @Override
+        protected AsyncShardFetch.FetchResult<NodeGatewayStartedShards>
+        fetchData(ShardRouting shard, RoutingAllocation allocation) {
+            // for now always return immediately what we know
+            final ShardId shardId = shard.shardId();
+            final Set<String> ignoreNodes = allocation.getIgnoreNodes(shardId);
+            Map<DiscoveryNode, NodeGatewayStartedShards> foundShards = knownAllocations.values().stream()
+                .flatMap(shardMap -> shardMap.values().stream())
+                .filter(ks -> ks.shardId().equals(shardId))
+                .filter(ks -> ignoreNodes.contains(ks.currentNodeId()) == false)
+                .filter(ks -> knownNodes.containsKey(ks.currentNodeId()))
+                .collect(Collectors.toMap(
+                    routing -> knownNodes.get(routing.currentNodeId()),
+                    routing ->
+                        new NodeGatewayStartedShards(
+                            knownNodes.get(routing.currentNodeId()), -1, routing.allocationId().getId(), routing.primary())));
+
+            return new AsyncShardFetch.FetchResult<>(shardId, foundShards, Collections.emptySet(), ignoreNodes);
+        }
+    };
+
+    ReplicaShardAllocator replicaShardAllocator = new ReplicaShardAllocator(Settings.EMPTY) {
+        @Override
+        protected AsyncShardFetch.FetchResult<NodeStoreFilesMetaData> fetchData(ShardRouting shard, RoutingAllocation allocation) {
+            // for now, just pretend no node has data
+            final ShardId shardId = shard.shardId();
+            return new AsyncShardFetch.FetchResult<>(shardId, Collections.emptyMap(), Collections.emptySet(),
+                allocation.getIgnoreNodes(shardId));
+        }
+    };
+
+    public TestGatewayAllocator() {
```

Comment:
it is the first method. the rest are fields. I personally don't care. let me know what you prefer.